src/connection.c

```diff
@@ -0,0 +1,94 @@
+/* ==========================================================================
+ * connection.c - connection layer framework
+ * --------------------------------------------------------------------------
+ * Copyright (C) 2022  zhenwei pi
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the
+ * following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+ * NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+ * USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * ==========================================================================
+ */
+
+#include "server.h"
+#include "connection.h"
+
+static ConnectionType *connTypes[CONN_TYPE_MAX];
+
+int connTypeRegister(ConnectionType *ct) {
+    int type = ct->get_type(NULL);
+
+    /* unknown connection type as a fatal error */
+    if (type >= CONN_TYPE_MAX) {
+        serverPanic("Unsupported connection type %d", type);
+    }
+
+    if (connTypes[type] == ct) {
+        serverLog(LL_WARNING, "Connection type %d already registered", type);
+        return C_OK;
+    }
+
+    serverLog(LL_VERBOSE, "Connection type %d registered", type);
+    connTypes[type] = ct;
+
+    if (ct->init) {
+        ct->init();
+    }
+
+    return C_OK;
+}
+
+int connTypeInitialize() {
+    /* currently socket connection type is necessary  */
+    serverAssert(RedisRegisterConnectionTypeSocket() == C_OK);
+
+    /* may fail if without BUILD_TLS=yes */
+    RedisRegisterConnectionTypeTLS();
+
+    return C_OK;
+}
+
+ConnectionType *connectionByType(int type) {
+    ConnectionType *ct;
+
+    serverAssert(type < CONN_TYPE_MAX);
+
+    ct = connTypes[type];
+    if (!ct) {
+        serverLog(LL_WARNING, "Missing implement of connection type %d", type);
```

Comment:
I realized that this chunk of change is a hard work to review, so I decided to split this into 3 steps:
- In this version, the TLS still works statically, and registering types into an array.
- registering connection types into a lookup table with their names(string) and making TLS dynamically loadable.
- implementing RDMA connection type.

I'm OK to change plan, using a lookup table in this series is good to me too. Please let me know your decision.  :D

Reply:
i took this note early in the review, before looking at the plan and commits in the other branch.
i agree with your approach of doing it gradually, and i don't mind to keep printing an integer for now.
i suppose we can even panic in this specific case.

p.s. when we'll go dynamic, we can still use an array (with max of 8 entries), each type will be able to report it's name with a method returning it.
i think we may be able to avoid any direct / hard-coded indexing into that array, except for maybe TLS for which we can just reserve index 1.
other extensions can be added in a way that redis doesn't need to explicitly refer to them (they'll register their own configs, etc), and the only one that will be hard-wired into redis it TLS (in order to be backwards compatible with the old configs).